Source/WebCore/css/StyleProperties.cpp

```diff
@@ -685,16 +650,44 @@ String StyleProperties::textDecorationSkipValue() const
 
 String StyleProperties::fontVariantValue() const
 {
-    String commonValue;
     StringBuilder result;
-    appendFontLonghandValueIfExplicit(CSSPropertyFontVariantLigatures, result, commonValue);
-    if (isCSSWideValueKeyword(result.toString()))
-        return result.toString();
-    appendFontLonghandValueIfExplicit(CSSPropertyFontVariantCaps, result, commonValue);
-    appendFontLonghandValueIfExplicit(CSSPropertyFontVariantAlternates, result, commonValue);
-    appendFontLonghandValueIfExplicit(CSSPropertyFontVariantNumeric, result, commonValue);
-    appendFontLonghandValueIfExplicit(CSSPropertyFontVariantEastAsian, result, commonValue);
-    appendFontLonghandValueIfExplicit(CSSPropertyFontVariantPosition, result, commonValue);
+    std::optional<CSSValueID> commonKeyword;
+    bool isAllNormal = true;
+
+    for (auto property : fontVariantShorthand()) {
+        int foundPropertyIndex = findPropertyIndex(property);
+        if (foundPropertyIndex == -1)
+            return emptyString(); // All longhands must have at least implicit values.
+
+        auto value = propertyAt(foundPropertyIndex).value();
+        auto keyword = valueID(value);
+
+        // Skip normal for brevity.
+        if (keyword == CSSValueNormal)
+            continue;
+
+        isAllNormal = false;
+
+        // If all properties are set to the same special keyword, serialize as that.
+        // If some but not all properties are, the font shorthand can't represent that, serialize as empty string.
+        if (isCSSWideKeyword(keyword) || CSSPropertyParserHelpers::isSystemFontShorthand(keyword)) {
```

Comment:
This `isSystemFontShorthand` check is not quite right. Unlike the `fontValue` function, the `fontVariantValue` function should **never** return a system font name, instead should serialize as an empty string if any of the longhands have the system font name. We should add a test that checks this if we don’t already have one.

Reply:
Changed to return an empty string. Though we have the same problem with serializing font-family/size, as one of the WPT shows.